pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,353 @@
+package restrictusers
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/labels"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	"github.com/openshift/origin/pkg/controller/shared"
+)
+
+const (
+	// CollaboratorAnnotation is the key for an annotation on Namespace that
+	// specifies collaborators on the namespace.
+	CollaboratorAnnotation = "authorization.openshift.io/collaborators"
+)
+
+// CollaboratorType is a type of collaborator on a project.
+type CollaboratorType string
+
+const (
+	// CollaboratorTypeUser indicates a collaborator that is a regular user.
+	CollaboratorTypeUser CollaboratorType = "User"
+
+	// CollaboratorTypeUser indicates a collaborator that is a regular group.
+	CollaboratorTypeGroup CollaboratorType = "Group"
+
+	// CollaboratorTypeServiceAccount indicates a collaborator that is a service
+	// account.
+	CollaboratorTypeServiceAccount CollaboratorType = "ServiceAccount"
+)
+
+// CollaboratorMatcher is an object that can be matched against a subject (user,
+// group, or service account) to determine if that subject is a collaborator on
+// the namespace to which the CollaboratorMatcher belongs. A namespace has an
+// array of CollaboratorMatcher objects, stored as serialized as JSON in an
+// annotation with the key CollaboratorAnnotation.  If any one of those
+// CollaboratorMatcher objects matches a subject, that subject is considered to
+// be a collaborator on the namespace.  Each CollaboratorMatcher object
+// specifies a type of collaborator (CollaboratorType) and either a list of
+// literal names or a label query (in which case the list must be empty).
+type CollaboratorMatcher struct {
+	unversioned.TypeMeta `json:",inline"`
+	kapi.ObjectMeta      `json:"metadata,omitempty"`
+
+	// Type is the type of collaborators against which this CollaboratorMatcher
+	// may be matched.
+	Type CollaboratorType `json:"type"`
```

Comment:
Wrap these and call them `Spec`. I can envision a world where this controller driven and a controller drives updates to status to describe matches.  That would prevent multiple lookup/matching calls on every API server and allow an admin to see who actually matches for something like content assist from a `Status` field.